clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -455,21 +476,31 @@ public void onSuccess(ByteBuffer value) {
                     // handle join completion in the callback so that the callback will be invoked
                     // even if the consumer is woken up before finishing the rebalance
                     synchronized (AbstractCoordinator.this) {
-                        log.info("Successfully joined group with generation {}", generation.generationId);
-                        state = MemberState.STABLE;
-                        rejoinNeeded = false;
-                        // record rebalance latency
-                        lastRebalanceEndMs = time.milliseconds();
-                        sensors.successfulRebalanceSensor.record(lastRebalanceEndMs - lastRebalanceStartMs);
-                        lastRebalanceStartMs = -1L;
-
-                        if (heartbeatThread != null)
-                            heartbeatThread.enable();
+                        // Generation data maybe concurrently cleared by Heartbeat thread.
+                        if (generation.generationId != OffsetCommitRequest.DEFAULT_GENERATION_ID) {
+                            log.info("Successfully joined group with generation {}", generation.generationId);
+                            state = MemberState.STABLE;
+                            rejoinNeeded = false;
+                            // record rebalance latency
+                            lastRebalanceEndMs = time.milliseconds();
+                            sensors.successfulRebalanceSensor.record(lastRebalanceEndMs - lastRebalanceStartMs);
+                            lastRebalanceStartMs = -1L;
+
+                            if (heartbeatThread != null)
+                                heartbeatThread.enable();
+                        } else {
+                            log.info("Generation data was cleared by heartbeat thread. Rejoin failed.");
+                            failure();
+                        }
                     }
                 }
 
                 @Override
                 public void onFailure(RuntimeException e) {
+                    failure();
+                }
+
+                private void failure() {
```

Comment:
Having the synchronized block inside `failure` means that we would lock twice on `AbstractCoordinator.this`. I'd suggest we leave the locking at the caller can just leave `failure` without it.